cloudferrylib/utils/retrying.py

```diff
@@ -0,0 +1,199 @@
+# limitations under the License.
+
+import sys
+import time
+from functools import wraps
+
+from cloudferrylib.utils import log
+
+LOG = log.getLogger(__name__)
+
+
+UNLIMITED = None
+
+
+class RetryError(RuntimeError):
+    pass
+
+
+class RetryRequired(RetryError):
+    pass
+
+
+class TimeoutExceeded(RetryError):
+    pass
+
+
+class MaxAttemptsReached(RetryError):
+    pass
+
+
+class Retry(object):
+    """Retries method calls based on configuration provided
+
+     :param max_attempts: maximum number of attempts to retry
+     :param max_time: maximum time to wait
+     :param wait_exponential: sleep time will grow exponentially if set to
+     true, linear otherwise. Exponent factor is 2.
+     :param raise_error: when set to True - raises error if all attempts
+     failed otherwise continues silently
+     :param reraise_original_exception: if set to True, re-raises last
+     exception of function instead of MaxAttemptsReached or TimeoutException.
+     :param retry_on_return_value: retries if return value matches to
+     `return_value`
+     :param return_value: see :param retry_on_return_value:
+     :param timeout: sleep timeout when :param wait_exponential: is disabled
+     :param expected_exceptions: list of exceptions which will be ignored
+     and raised to user without sleep
+     :param predicate: predicate to check after each function call
+     :param predicate_args: predicate arguments
+     :param predicate_kwargs: predicate key-value arguments
+     :param predicate_retval_as_arg: use function's return value as argument
+     to predicate
+     :param retry_message: message to be printed into log on each failure
+     :param stop_on_return_value: retries are stopped when return value matches
+    """
+
+    def __init__(self,
+                 max_attempts=1,
+                 max_time=UNLIMITED,
+                 wait_exponential=True,
+                 raise_error=True,
+                 reraise_original_exception=False,
+                 retry_on_return_value=False,
+                 return_value=None,
+                 timeout=1,
+                 expected_exceptions=None,
+                 predicate=None,
+                 predicate_args=None,
+                 predicate_kwargs=None,
+                 predicate_retval_as_arg=False,
+                 retry_message=None,
+                 stop_on_return_value=False,
+                 print_stack_trace=True):
+        self.print_stack_trace = print_stack_trace
+        self.reraise_original_exception = reraise_original_exception
+        self.timeout = timeout
+        self.total_time = 0
+        self.attempt = 0
+        self.max_attempts = max_attempts
+        self.wait_exponential = wait_exponential
+        self.max_time = max_time
+        self.raise_error = raise_error
+        self.retry_on_return_value = retry_on_return_value
+        self.return_value = return_value
+        self.expected_exceptions = expected_exceptions or []
```

Comment:
Consider using tuple to store expected exceptions, like `self.expected_exceptions = tuple(expected_exceptions or [])`, this way you will be able to use `isinstance` without loop, e.g. `isinstance(obj, self.expected_exceptions)`